### PR TITLE
mgr/dasboard_v2: Refactor controller

### PR DESCRIPTION
**Pull request by votdev**
Refactor the controller implementation to make it a little bit more OO. This way the ``mgr`` and ``logger`` properties are well designed and documented which should help contributors to understand the code.

Signed-off-by: Volker Theile <vtheile@suse.com>

### CODE REVIEW

**Review by sebastian-philipp — APPROVED** [review]

File: src/pybind/mgr/dashboard_v2/tools.py [hunk 1]
```diff
@@ -59,7 +60,6 @@ def load_controllers(mgrmodule):
             if isinstance(cls, type) and hasattr(cls, '_cp_controller_'):
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/tools.py]: This must be secured. Check for BaseController.

File: src/pybind/mgr/dashboard_v2/tools.py [hunk 2]
```diff
@@ -70,7 +70,51 @@ def _json_error_page(status, message, traceback, version):
...
+
+
+class BaseController(six.with_metaclass(BaseControllerMeta, object)):
+    """
+    Base class for all controllers providing API endpoints.
+    """
+    _mgr_module = None
+
+    @property
+    def mgr(self):
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/tools.py:103]: I don't think we need to also re-declare this property here, I think `self.mgr` will be directed to the class method. But you need to confirm this.

**Review by rjfd — APPROVED** [review]
lgtm, good idea!

File: src/pybind/mgr/dashboard_v2/tools.py (hunk 2 above)
Comment (votdev) [reply src/pybind/mgr/dashboard_v2/tools.py:103]: Found out the instance properties are needed.

**Review by LenzGr — APPROVED** [review]

File: src/pybind/mgr/dashboard_v2/tools.py [hunk 3]
```diff
@@ -56,10 +57,9 @@ def load_controllers(mgrmodule):
         mod = importlib.import_module('.controllers.{}'.format(mod_name),
                                       package='dashboard_v2')
         for _, cls in mod.__dict__.items():
-            if isinstance(cls, type) and hasattr(cls, '_cp_controller_'):
-                # found controller
+            # Controllers MUST be derived from the class BaseController.
+            if isinstance(cls, BaseControllerMeta) and hasattr(cls, '_cp_controller_'):
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/tools.py:61]: `isinstance(cls, BaseControllerMeta)` or `isinstance(cls, BaseController)` ?

**Comment by rjfd** [comment]
@votdev please update the developer notes in the README to say that the controller should extend from the `BaseController` class

File: src/pybind/mgr/dashboard_v2/tools.py (hunk 3 above)
Comment (votdev) [reply src/pybind/mgr/dashboard_v2/tools.py:61]: isinstance(cls, BaseControllerMeta) because type(cls) returns ``BaseControllerMeta``.

**Comment by rjfd** [comment]
@votdev also, add the `mgr/dashboard_v2` prefix to the commit messages

**Review by jecluis — CHANGES_REQUESTED** [review]
commit header needs changing:

1. component to be added to the commit title;
2. commit title to be shorter;
3. instead of having a lot of information in the title, maybe move most of it to the commit's body.

**Review by p-se — APPROVED** [review]

**Review by jecluis — CHANGES_REQUESTED** [review]
@votdev I did not mean the PR's title (although that is also appreciated!). The actual patch commit is what really needs changing before we merge it.

**Comment by votdev** [comment]
@rjfd README.rst has been adapted.

**Comment by votdev** [comment]
@jecluis Commit title has been adapted

**Comment by jecluis** [comment]
@votdev thanks!

**Review by jecluis — APPROVED** [review]

**Pull request merged**